src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/DeserializationHelpers.cs

```diff
@@ -57,6 +58,36 @@ internal static class DeserializationHelpers
             }
         }
 
+
+        internal static void ApplyInstanceAnnotations(object resource, IEdmStructuredTypeReference structuredType, ODataResource oDataResource, 
+            ODataDeserializerProvider deserializerProvider, ODataDeserializerContext readContext)
+        {   
+            PropertyInfo propertyInfo = EdmLibHelpers.GetInstanceAnnotationsDictionary(structuredType.StructuredDefinition(), readContext.Model);
+            if (propertyInfo == null)
+            {
+                return;
+            }
+
+            IDelta delta = resource as IDelta;
+            object value;
+            if (delta != null)
+            {                
+                delta.TryGetPropertyValue(propertyInfo.Name, out value);
+            }
+            else
+            {
+                value = propertyInfo.GetValue(resource);
+            }
+
+            IDictionary<string, IDictionary<string, object>> instAnnotationDictionary = value as IDictionary<string, IDictionary<string, object>>;
+            if (instAnnotationDictionary == null)
+            {
+                instAnnotationDictionary = new Dictionary<string, IDictionary<string, object>>();
```

Comment:
>instAnnotationDictionary = new Dictionary<string, IDictionary<string, object>>(); [](start = 16, length = 81)

We should set the instanceAnnotationDictionary on the resource here, rather than in SetInstanceAnnotations, so we don't have to handle this special case in SetInstanceAnnotations (and don't need to pass the resource or propertyInfo.Name)